### PR TITLE
stdenv: run patchShebangs on the configure script when it's a file

### PR DESCRIPTION
**Pull request by Artturin**
if the configure script has a `/usr/bin/env` or some other shebang which is not in the sandbox then there will be errors such as

`...-stdenv-linux/setup: line 1299: ./configure: cannot execute: required file not found`

There are 250 files which `patchShebangs` `./configure`

https://github.com/search?q=NOT+is%3Afork+lang%3Anix+%2FpatchShebangs+.%5C%2Fconfigure%2F&type=code

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
Rebuilding the world with this; should be done in ~10 hours.

**Comment by Artturin** [comment]
> Rebuilding the world with this; should be done in ~10 hours.

:+1: , I still have to add the treewide removal of the now unnecessary patchShebang's but

**Comment by Artturin** [comment]
Resolved conflict

**Review by trofi — APPROVED** [review]
LGTM.

You might also want to wrap it into a `if [ -z "${dontPatchConfigure:-}" ]; then` in an unlikely case when it does some harm (I can't come up with a plausible scenario though).

**Review by ghost — APPROVED** [review]
I agree with @trofi that

> `if [ -z "${dontPatchConfigure:-}" ]; then`

is a good idea.  Adding it now means the first person who finds a need for it won't have to cause a mass-rebuild.

**Review by trofi — APPROVED** [review]

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -1273,6 +1273,10 @@ configurePhase() {
                 prependToVar configureFlags --disable-static
             fi
         fi
+
+        if [ -z "${dontPatchShebangsInConfigure:-}" ]; then
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1277]: Changed to be more specific

**Review by trofi — APPROVED** [review]

**Pull request merged**